src/storm-pomdp/parser/PomdpSolveParser.cpp

```diff
@@ -0,0 +1,422 @@
+    std::vector<double> start_probabilities;
+    std::map<std::string, std::map<std::string, double>> transitions;
+    std::map<std::string, std::map<std::string, double>> observations_prob;
+    std::map<std::string, double> rewards;
+    std::vector<std::string> newStates; // Stores unique new states with observation incorporated
+    std::map<std::string, std::map<std::string, double>> newTransitions; // Modified transitions with observation-based states
+
+    std::map<std::string, std::map<std::string, double>> createNewTransitions() {
+        std::unordered_set<std::string> uniqueStates;
+
+        for (const auto& action_entry : transitions) {
+            const std::string& action = action_entry.first;
+
+            for (const auto& transition_entry : action_entry.second) {
+                auto pos = transition_entry.first.find(':');
+                std::string startState = transition_entry.first.substr(0, pos);
+                std::string endState = transition_entry.first.substr(pos + 1);
+                double transition_prob = transition_entry.second;
+
+                double totalProbability = 0.0;
+
+                if (observations_prob.count(action) > 0) {
+                    std::string obs_key_prefix = endState + ":";
+                    for (const auto& obs_entry : observations_prob.at(action)) {
+                        if (obs_entry.first.rfind(obs_key_prefix, 0) == 0) {
+                            std::string observation = obs_entry.first.substr(obs_key_prefix.length());
+                            double obs_prob = obs_entry.second;
+
+                            double new_prob = transition_prob * obs_prob;
+                            std::string newEndState = endState + ":" + observation;
+
+                            if (new_prob > 0) {
+                                std::string actionEndStateKey = action + ":" + newEndState;
+                                newTransitions[startState][actionEndStateKey] = new_prob;
+                                totalProbability += new_prob;
+                                uniqueStates.insert(newEndState);
+                            }
+                        }
+                    }
+                }
+
+                if (totalProbability < transition_prob) {
+                    double remaining_prob = transition_prob - totalProbability;
+                    std::string actionEndStateKey = action + ":" + endState;
+                    newTransitions[startState][actionEndStateKey] = remaining_prob;
+                }
+            }
+        }
+
+        newStates.assign(uniqueStates.begin(), uniqueStates.end());
+        return newTransitions;
+    }
+
+    void addInitialStateTransitions() {
+        const std::string initialState = "initial";
+        const std::string action = "start";
+
+        if (start_probabilities.size() != states.size()) {
+            std::cerr << "Error: start_probabilities and states vectors must be of the same size." << std::endl;
```

Comment:
For logging, warnings and errors, there are macros STORM_PRINT_AND_LOG, STORM_LOG_WARN, STORM_LOG_ERROR, ...; please use those for output if possible